riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+
+        // This fits in 19 bits.
+        //XX = (tmp3_h - tmp5_h)**2 + (tmp3_l - tmp5_l)**2 + (tmp4_h - tmp6_h)**2 + (tmp4_l - tmp6_l)**2,
+        XX = (tmp3_h - tmp5_h)*(tmp3_h - tmp5_h) + (tmp3_l - tmp5_l)*(tmp3_l - tmp5_l) + (tmp4_h - tmp6_h)*(tmp4_h - tmp6_h) + (tmp4_l - tmp6_l)*(tmp4_l - tmp6_l),
+
+        pc' = (1 - XXIsZero) * l + XXIsZero * (pc + 1)
+    }
+
+    // Jump to `l` if (val(X) - val(Y)) == Z, where X and Y are register ids and Z is a number.
+    instr branch_if_diff_equal XL, YL, ZH, ZL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = add_sub.sub(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> (tmp4_h, tmp4_l) = add_sub.sub(tmp3_h, tmp3_l, ZH, ZL)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = tmp4_h + tmp4_l,
+        pc' = XXIsZero * l + (1 - XXIsZero) * (pc + 1)
+    }
+
+    // Skips W instructions if val(X) - val(Y) + Z is zero, where X and Y are register ids and Z is a
+    // constant offset.
+    instr skip_if_equal XL, YL, ZH, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = add_sub.sub(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> (tmp4_h, tmp4_l) = add_sub.add(tmp3_h, tmp3_l, ZH, ZL)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = tmp4_h + tmp4_l,
+        pc' = pc + 1 + (XXIsZero * WL)
+    }
+
+    // Branches to `l` if val(X) >= val(Y) <=> not(val(X) < val(Y)).
+    instr branch_if_greater_or_equal XL, YL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> wrap_bit = add_sub.gt(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+    {
+        pc' = (1 - wrap_bit) * l + wrap_bit * (pc + 1)
+    }
+
+    // Branches to `l` if val(X) >= val(Y) <=> not(val(X) < val(Y)).
+    instr branch_if_greater_or_equal_signed XL, YL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> tmp5_h = add_sub.gt(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link => byte.check(X_b1)
+        link => byte.check(X_b2)
+        link => byte.check(X_b3)
+        link => bit7.check(tmp3_l)
+        link => bit7.check(tmp3_h)
+    {
+        // TODO witgen doesn't work with this yet.
+        std::utils::force_bool(tmp5_h),
+
+        tmp1_h = X_b1 + tmp3_l * 2**8 + wrap_bit * 2**15,
+        tmp2_h = X_b2 + tmp3_h * 2**8 + wrap_bit_2 * 2**15,
+
+        tmp4_l = (1 - wrap_bit) * (1 - wrap_bit_2),
+        tmp4_h = wrap_bit * wrap_bit_2,
+        tmp5_l = (1 - wrap_bit) * wrap_bit_2,
+        //... = wrap_bit * (1 - wrap_bit_2),
+
+        X_b3 = tmp4_l * (1 - tmp5_h) +
+               tmp4_h * (1 - tmp5_h) +
+               tmp5_l, /* * 1 */
+               /* ...* 0, */
```

Comment:
This would be much nicer if we could have "aliases" for columns inside instructions.
The idea is as follows:
We need to compare two signed numbers: `wrap_bit` and `wrap_bit_2` are the signs.
`tmp4_l`: are both numbers positive?
`tmp4_h`: are both numbers negative?
`tmp5_l`: is the first number positive and the second number negative?

`X_b3` does the math for whether we should jump or not (i.e., X >= Y), considering the different combinations of signs and the result of the comparison operation in the submachine (add_sub.gt)

Reply:
What about the commented out parts? What's the reasoning there?

Reply:
They were left as documentation to show the full truth table, but not needed cus mul by 0

Reply:
added some text comments to clarify